### PR TITLE
Enhance Edit Command with Remark and Tag Fields

### PR DESCRIPTION
**Pull request by wapisai**
## This PR addresses the issues #121 #99 
This Pull Request introduces the ability to edit remark and tag fields using the Edit command.

## Changes Made:
- Edit Command Enhancement: The Edit command has been updated to allow modifications to remark and tag fields, offering users enhanced control over their data.
- Unit Test Expansion: Introduced comprehensive unit tests in EditCommandTest and EditCommandParserTest to rigorously assess the new functionalities. These tests cover a wide array of scenarios, including both edge cases and common use patterns, to ensure the feature's reliability.

## Testing:
- Unit Testing: Developed and executed a suite of new unit tests designed to rigorously validate the edit operation across various scenarios, emphasizing error handling and behavior correctness.
- Manual Testing: Performed extensive manual testing to guarantee that the integration of this new feature does not disrupt existing functionalities, ensuring a seamless user experience.

## Code Quality and Compatibility:
- Compatibility Assurance: Verified the new feature's compatibility with existing commands and functionalities through detailed automated regression testing, ensuring no disruptions to the current system.
- Code Quality Maintenance: Conducted thorough code quality checks to uphold our commitment to high standards and adherence to the project's coding conventions.

## Additional Notes:
Priority for this feature has been set to Medium, reflecting its importance to user experience.
We encourage further review and testing from the community to refine and validate the new tests and functionalities introduced.

## Please feel free to provide feedback or suggestions for further improvements!

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DeleteCommand.java [hunk 1]
```diff
@@ -19,31 +22,47 @@ public class DeleteCommand extends Command {
...
         requireNonNull(model);
         List<Person> lastShownList = model.getFilteredPersonList();
+        Person personToDelete;
 
-        if (targetIndex.getZeroBased() >= lastShownList.size()) {
-            throw new CommandException(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        if (targetIndex != null) {
+            personToDelete = DeleteCommandExecutor.deleteByIndex(lastShownList, targetIndex);
+        } else {
+            personToDelete = DeleteCommandExecutor.deleteByName(lastShownList, targetName);
```
Comment (appleraincoat) [inline src/main/java/seedu/address/logic/commands/DeleteCommand.java:63]: I like your use of SLAP!

File: src/main/java/seedu/address/logic/commands/delete/DeleteCommandExecutor.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Contains the logic for executing the delete command.
+ */
+public class DeleteCommandExecutor {
+
+    /**
+     * Deletes a client from the address book using their index in the last shown list.
+     */
+    public static Person deleteByIndex(List<Person> lastShownList, Index targetIndex) throws CommandException {
+        if (targetIndex.getZeroBased() >= lastShownList.size()) {
```
Comment (appleraincoat) [inline src/main/java/seedu/address/logic/commands/delete/DeleteCommandExecutor.java:20]:
```suggestion
        if (targetIndex.getZeroBased() >= lastShownList.size() || targetIndex.getZeroBased() < 1) {
```

Comment (appleraincoat) [reply src/main/java/seedu/address/logic/commands/delete/DeleteCommandExecutor.java:20]: Perhaps you can consider adding this extra check, which also includes the case where 0 or negative numbers are passed in, which is invalid as well.
Similar for EditCommand line 85

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 3]
```diff
@@ -68,28 +63,16 @@ public EditCommand parse(String args) throws ParseException {
...
             editPersonDescriptor.setFamily(ParserUtil.parseFamily(argMultimap.getValue(PREFIX_FAMILY).get()));
         }
-        parseTagsForEdit(argMultimap.getAllValues(PREFIX_TAG)).ifPresent(editPersonDescriptor::setTags);
-
+        if (argMultimap.getValue(PREFIX_TAG).isPresent()) {
+            editPersonDescriptor.setTags(ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG)));
+        }
+        if (argMultimap.getValue(PREFIX_REMARK).isPresent()) {
+            editPersonDescriptor.setRemark(ParserUtil.parseRemark(argMultimap.getValue(PREFIX_REMARK).get()));
+        }
```
Comment (appleraincoat) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java:71]: 👍

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 4]
```diff
@@ -41,6 +42,20 @@ public void execute_validIndexUnfilteredList_success() {
...
+
+        String expectedMessage = String.format(DeleteCommand.MESSAGE_DELETE_PERSON_SUCCESS,
+                Messages.format(personToDelete));
+
+        ModelManager expectedModel = new ModelManager(model.getAddressBook(), new UserPrefs());
+        expectedModel.deletePerson(personToDelete);
+
+        assertCommandSuccess(deleteCommand, model, expectedMessage, expectedModel);
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:58]: Good test for testing successful deletion by name.

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 5]
```diff
@@ -49,6 +64,14 @@ public void execute_invalidIndexUnfilteredList_throwsCommandException() {
...
     }
 
+    @Test
+    public void execute_invalidNameUnfilteredList_throwsCommandException() {
+        Person personToDelete = new PersonBuilder().withName("b").build();
+        DeleteCommand deleteCommand = new DeleteCommand(personToDelete.getName());
+
+        assertCommandFailure(deleteCommand, model, Messages.MESSAGE_INVALID_PERSON_DISPLAYED_NAME);
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:74]: Good test for testing unsuccessful deletion of name that does not exist.

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 6]
```diff
@@ -101,6 +124,41 @@ public void equals() {
...
     }
 
+    @Test
+    public void equalsName() {
+        DeleteCommand deleteUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya").build().getName());
+        DeleteCommand deleteNotUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("not Udhaya").build().getName());
+        DeleteCommand deleteUdhayaShanugamCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya Shanugam").build().getName());
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:134]:
```suggestion
                new PersonBuilder().withName("Udhaya Shanmugam").build().getName());
```

Comment (appleraincoat) [reply src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:134]: 😢

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 7]
```diff
@@ -101,6 +124,41 @@ public void equals() {
         assertFalse(deleteFirstCommand.equals(deleteSecondCommand));
     }
 
+    @Test
+    public void equalsName() {
+        DeleteCommand deleteUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("Udhaya").build().getName());
+        DeleteCommand deleteNotUdhayaCommand = new DeleteCommand(
+                new PersonBuilder().withName("not Udhaya").build().getName());
+        DeleteCommand deleteUdhayaShanugamCommand = new DeleteCommand(
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:133]:
```suggestion
        DeleteCommand deleteUdhayaShanmugamCommand = new DeleteCommand(
```

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 8]
```diff
@@ -101,6 +124,41 @@ public void equals() {
...
+        assertFalse(deleteUdhayaCommand.equals(null));
+
+        // different person -> returns false
+        assertFalse(deleteUdhayaCommand.equals(deleteNotUdhayaCommand));
+
+        // delete udhaya should not delete someone with address udhaya
+        assertFalse(deleteUdhayaCommand.equals(deleteAddressCommand));
+
+        //delete Udhaya should not delete someone with name Udhaya Shanugam
+        assertFalse(deleteUdhayaCommand.equals(deleteUdhayaShanugamCommand));
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:159]:
```suggestion
        assertFalse(deleteUdhayaCommand.equals(deleteUdhayaShanmugamCommand));
```

File: src/test/java/seedu/address/logic/parser/DeleteCommandParserTest.java [hunk 9]
```diff
@@ -27,6 +27,6 @@ public void parse_validArgs_returnsDeleteCommand() {
 
     @Test
     public void parse_invalidArgs_throwsParseException() {
-        assertParseFailure(parser, "a", String.format(MESSAGE_INVALID_COMMAND_FORMAT, DeleteCommand.MESSAGE_USAGE));
+        assertParseFailure(parser, "a", String.format(MESSAGE_INVALID_INDEX, DeleteCommand.MESSAGE_USAGE));
```
Comment (appleraincoat) [inline src/test/java/seedu/address/logic/parser/DeleteCommandParserTest.java:30]: 👍

**Review by appleraincoat — COMMENTED** [review]
Great job!
Remember to update the extra checks for index validity for deletecommand and editcommand

**Comment by 4llysa** [comment]
Currently edit does not support editing tags to include more than one tag. For eg, if a client was initially just a seller but now also wants to buy a property and the real estate agent wishes to also tag the client as both buyer and seller. 

We can update this feature in v1.3 perhaps.

**Review by appleraincoat — APPROVED** [review]
LGTM!

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -335,30 +335,30 @@ Priorities: High (must have) - `* * *`, Medium (nice to have) - `* *`, Low (unli
...
     * 1e1. Realodex throws an error and highlights the format to user.
-    * 1e2. User enters new data. 
+    * 1e2. User enters new data.
     * Use case resumes from step 1.
 
-* 1f. Address is not fully English  
+* 1f. Address is not fully English
     * 1f1. Realodex throws an error and requests for only English input.
-    * 1f2. User enters new data. 
+    * 1f2. User enters new data.
```
Comment (appleraincoat) [inline docs/DeveloperGuide.md:361]: Thank you for updating the DG! LGTM.

File: src/main/java/seedu/address/logic/commands/delete/DeleteCommandExecutor.java [hunk 11]
```diff
@@ -17,7 +17,7 @@ public class DeleteCommandExecutor {
      * Deletes a client from the address book using their index in the last shown list.
      */
     public static Person deleteByIndex(List<Person> lastShownList, Index targetIndex) throws CommandException {
-        if (targetIndex.getZeroBased() >= lastShownList.size()) {
+        if (targetIndex.getZeroBased() >= lastShownList.size() || targetIndex.getZeroBased() < 0) {
```
Comment (appleraincoat) [inline src/main/java/seedu/address/logic/commands/delete/DeleteCommandExecutor.java:20]: 👍

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java [hunk 12]
```diff
@@ -102,6 +103,15 @@ public EditPersonDescriptorBuilder withTags(String... tags) {
...
 
+    /**
+     * Sets the {@code Remark} of the {@code EditPersonDescriptor} that we are building.
+     */
+    public EditPersonDescriptorBuilder withRemark(String remark) {
+        Remark r = new Remark(remark);
+        descriptor.setRemark(r);
+        return this;
+    }
+
```
Comment (appleraincoat) [inline src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java:114]: 👍

**Review by appleraincoat — APPROVED** [review]
LGTM!

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 13]
```diff
@@ -68,28 +63,16 @@ public EditCommand parse(String args) throws ParseException {
         if (argMultimap.getValue(PREFIX_FAMILY).isPresent()) {
             editPersonDescriptor.setFamily(ParserUtil.parseFamily(argMultimap.getValue(PREFIX_FAMILY).get()));
         }
-        parseTagsForEdit(argMultimap.getAllValues(PREFIX_TAG)).ifPresent(editPersonDescriptor::setTags);
-
+        if (argMultimap.getValue(PREFIX_TAG).isPresent()) {
+            editPersonDescriptor.setTags(ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG)));
+        }
+        if (argMultimap.getValue(PREFIX_REMARK).isPresent()) {
+            editPersonDescriptor.setRemark(ParserUtil.parseRemark(argMultimap.getValue(PREFIX_REMARK).get()));
```
Comment (jyztintan) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java:70]: I realised that if you had ever put a remark for the person, currently there is no functionality to clear the remark under edit?

meaning to say that once you put a remark for a client there must always be a remark for that person. 
not sure if this is something we might want to consider in future

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 14]
```diff
@@ -272,7 +274,8 @@ public boolean equals(Object other) {
                     && Objects.equals(email, otherEditPersonDescriptor.email)
                     && Objects.equals(address, otherEditPersonDescriptor.address)
                     && Objects.equals(family, otherEditPersonDescriptor.family)
-                    && Objects.equals(tags, otherEditPersonDescriptor.tags);
+                    && Objects.equals(tags, otherEditPersonDescriptor.tags)
+                    && Objects.equals(remark, otherEditPersonDescriptor.remark);
```
Comment (jyztintan) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:278]: LGTM!

File: src/test/java/seedu/address/logic/commands/DeleteCommandTest.java [hunk 15]
```diff
@@ -130,7 +130,8 @@ public void equalsName() {
                 new PersonBuilder().withName("Udhaya").build().getName());
         DeleteCommand deleteNotUdhayaCommand = new DeleteCommand(
                 new PersonBuilder().withName("not Udhaya").build().getName());
-        DeleteCommand deleteUdhayaShanugamCommand = new DeleteCommand(
+        DeleteCommand deleteUdhayaShanmugamCommand = new DeleteCommand(
```
Comment (jyztintan) [inline src/test/java/seedu/address/logic/commands/DeleteCommandTest.java:133]: Thank you for fixing this typo! 👍

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java (hunk 12 above)
Comment (jyztintan) [reply src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java:114]: Looks like good SLAP!

**Review by jyztintan — APPROVED** [review]
Thanks for adding the changes :)

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java (hunk 13 above)
Comment (jyztintan) [reply src/main/java/seedu/address/logic/parser/EditCommandParser.java:70]: Nvm it works well
